File: src/main/java/lan/dk/podcastserver/manager/worker/finder/YoutubeFinder.java
package lan.dk.podcastserver.manager.worker.finder;

import lan.dk.podcastserver.entity.Cover;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.exception.FindPodcastNotFoundException;
import lan.dk.podcastserver.service.HtmlService;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.io.IOException;

/**
 * Created by kevin on 22/02/15.
 */
@Service("YoutubeFinder")
public class YoutubeFinder implements Finder {

    private final Logger logger = LoggerFactory.getLogger(this.getClass());

    private HtmlService htmlService;

    @Autowired
    public YoutubeFinder(HtmlService htmlService) {
        this.htmlService = htmlService;
    }

    @Override
    public Podcast find(String url) throws FindPodcastNotFoundException {

        Document page;

        try {
            page = htmlService.connectWithDefault(url).get();
        } catch (IOException e) {
            logger.error("IOException :", e);
            throw new FindPodcastNotFoundException();
        }

        Podcast youtubePodcst = new Podcast()
                .setUrl(url)
                .setType("Youtube")
                .setTitle(getTitle(page))
                .setDescription(getDescription(page));

        youtubePodcst
                .setCover(getCover(page));

        return youtubePodcst;
    }

    private String getDescription(Document page) {
        Element elementWithExternalId = page.select("meta[name=description]").first();
        if (elementWithExternalId != null) {
            return elementWithExternalId.attr("content");
        }

        return "";
    }

    private String getTitle(Document page) {
        Element elementWithExternalId = page.select("meta[name=title]").first();
        if (elementWithExternalId != null) {
            return elementWithExternalId.attr("content");
        }

        return "";
    }

    private Cover getCover(Document page) {
        Element elementWithExternalId = page.select("img.channel-header-profile-image").first();
        if (elementWithExternalId != null) {
            return new Cover(elementWithExternalId.attr("src"), 200, 200);
        }

        return new Cover();
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/worker/youtube/YoutubeFinder.kt
package com.github.davinkevin.podcastserver.manager.worker.youtube

import com.github.davinkevin.podcastserver.entity.Cover
import com.github.davinkevin.podcastserver.entity.Podcast
import com.github.davinkevin.podcastserver.find.FindCoverInformation
import com.github.davinkevin.podcastserver.find.FindPodcastInformation
import com.github.davinkevin.podcastserver.manager.worker.Finder
import com.github.davinkevin.podcastserver.service.CoverInformation
import com.github.davinkevin.podcastserver.service.ImageService
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.bodyToMono
import reactor.core.publisher.Mono
import reactor.core.publisher.toMono
import java.io.ByteArrayInputStream
import java.net.URI

/**
 * Created by kevin on 22/02/15
 */
@Service
class YoutubeFinder(
        private val imageService: ImageService,
        private val wcb: WebClient.Builder
) : Finder {

    override fun findInformation(url: String) = wcb
            .baseUrl(url)
            .build()
            .get()
            .retrieve()
            .bodyToMono<String>()
            .map { ByteArrayInputStream(it.toByteArray(Charsets.UTF_8)) }
            .map { Jsoup.parse(it, "UTF-8", url) }
            .map { toFindPodcastInformation(it, url) }

    override fun find(url: String): Podcast = TODO("not required anymore")

    private fun toFindPodcastInformation(p: Document, anUrl: String) = FindPodcastInformation (
        url = URI(anUrl),
        type = "Youtube",
        title = p.meta("title"),
        description = p.meta("description"),
        cover = findCover(p)?.toFindCover()
    )

    private fun findCover(page: Document): CoverInformation? {
        val coverUrl = page
                .select("img.channel-header-profile-image")
                .attr("src")

        return  if(coverUrl.isEmpty()) null
                else imageService.fetchCoverInformation(coverUrl)
    }

    override fun compatibility(url: String?) = _compatibility(url)
}

private fun Document.meta(s: String) = this.select("meta[name=$s]").attr("content")
private fun CoverInformation.toFindCover() = FindCoverInformation(height, width, url)
